### ISSUE TITLE
`Variable.unchain` is not reliable if the creator has multiple outputs

### ISSUE THREAD

**Issue by toslunar** [chainer/chainer#6027]
The documentation of the method `Variable.unchain` is correct: 

> Deletes the reference to the creator of this variable.

This does not always stop the creator from collecting the grad of the variable.

* Code to reproduce
```
>>> import numpy as np; import chainer; import chainer.functions as F
>>> x = chainer.Variable(np.arange(4).astype('f').reshape(2, 2)); x
variable([[0., 1.],
          [2., 3.]])
>>> h0, h1 = F.split_axis(x, [1], axis=0); h0, h1
(variable([[0., 1.]]), variable([[2., 3.]]))
>>> y = F.sum(h0 + h1)
>>> h0.unchain()
>>> y.backward()
>>> x.grad, h0.grad, h1.grad
(array([[1., 1.],
       [1., 1.]], dtype=float32), None, None)
```

**Comment by Rishav1** [comment]
This issue arises as the backward pass gathers gradients [from all output nodes](https://github.com/chainer/chainer/blob/1130eea1fe018eb3b61d112cade283ea2ccf5731/chainer/variable.py#L1030-L1031), even from output nodes on whom unchain() have been called. In order to curb this behavior, I think we can zero out gradients in [this line](https://github.com/chainer/chainer/blob/1130eea1fe018eb3b61d112cade283ea2ccf5731/chainer/variable.py#L1031) (without poping from _grads_) for nodes that have no _creator_ (i.e. have been unchained).

**Comment by toslunar** [comment]
@Rishav1 Thanks for the investigation and the concrete suggestion.  Are you interested in making it a pull request?
- Because my example is artificial, could you show your example where the issue matters, if you found the issue somewhere else?  A realistic example would raise the priority of the issue.
- In fact, I'm waiting for PRs to fix other backprop issues are merged.  The PRs #6348, #5952, #5722 might be relevant.  `#5274` and `#5724` are probably not.  If I fix the issue, I'll work after all of these.

**Comment by Rishav1** [comment]
@toslunar Thanks for responding to my comment. Take a look at this example similar to yours.

```python
>>> import numpy as np; import chainer; import chainer.functions as F
>>> x = chainer.Variable(np.arange(4).astype('f').reshape(2, 2)); x
variable([[0., 1.],
          [2., 3.]])
>>> h0, h1 = F.split_axis(x, [1], axis=0); h0, h1
(variable([[0., 1.]]), variable([[2., 3.]]))
>>> y = F.sum(h0)
>>> z = F.sum(h1)
>>> h0.unchain()
>>> x.grad, h0.grad, h1.grad
(None, None, None)
>>> y.backward()
>>> x.grad, h0.grad, h1.grad
(None, array([[1., 1.]], dtype=float32), None)
>>> z.backward()
>>> x.grad, h0.grad, h1.grad
(array([[1., 1.],
       [1., 1.]], dtype=float32), None, None)
```
Calling backward() on y correctly accumulates the gradient at the root h0 which means unchain is not the issue. But on calling backward() on z, the backward pass on reaching SplitAxis pulls the gradients from both the node outputs h0 and h1 into the input x. As I mentioned, I believe this behavior can be corrected by ensuring that if the child node is unchained(like h0), instead of pulling the gradient, use a dummy zeroed out gradient of same dimension as the child node input. 

If you agree on the fix, I would be glad to raise a PR. If you have a better alternative for the solution (like perhaps adding a fix in SplitAxis) please let me know and I would raise a request accordingly.

**Comment by toslunar** [comment]
Thank you, @Rishav1.

Your example is more realistic than mine.  It's sad to say it doesn't prove the importance of this issue, because your example would run as expected if #5722 is merged (and the compatibility option is turned off).  If you found the issue from a code like your example, you might be interested in #5709.

**Comment by toslunar** [comment]
I welcome @Rishav1 to raise a PR.  Let me revert my previous comment.

If this issue is resolved soon, the fix will be available in version 6.x as a backport, because the issue is purely a bug.  I don't expect the same for #5709, which has a difficult status as non-negligible number of users rely on the current behavior.

**Comment by Rishav1** [comment]
@toslunar, I raised a PR and added a test that fails without the change.

**Comment by toslunar** [comment]
The issue is fixed by #6804.
